packages/entity-cache-adapter-redis/src/__integration-tests__/ShardedRedisCacheAdapter-integration-test.ts

```diff
@@ -0,0 +1,218 @@
+import RedisTestEntity from '../testfixtures/RedisTestEntity';
+import { createRedisIntegrationTestEntityCompanionProvider } from '../testfixtures/createRedisIntegrationTestEntityCompanionProvider';
+
+//       shardingSchemeVersion: 1,
+
+const ids = {
+  entity1: '7345999c-b97d-47af-873c-bb1e9c1876f3', // shard 0
+  nonexistent: 'c207c238-70b5-41bb-8f33-130a57a6f6c5', // shard 1
+};
+
+function getShardForKey(key: string): number {
+  if (key.includes(ids.entity1)) {
+    return 0;
+  }
+
+  if (key.includes(ids.nonexistent)) {
+    return 1;
+  }
+
+  // entity1
+  if (key.includes('name:blah')) {
+    return 1; // put other field cache for entity1 on different server than the id field cached value
+  }
+
+  throw Error('unknown id encountered: ' + key);
+}
+
+function assertNotUndefined<T>(value: T | undefined): T {
+  invariant(value !== undefined, 'must not be undefined');
+  return value;
+}
+
+type SetInstruction = { key: string; value: string; secondsToken: 'EX'; seconds: number };
+
+class ShardedRedis implements IRedis {
+  constructor(private readonly redium: readonly Redis[]) {}
+
+  async mget(...args: string[]): Promise<(string | null)[]> {
+    const originalKeys = args;
+    const redisClientsForKeys = this.getRedisClientsForKeys(args);
+    const allKeysToValues = new Map<string, string | null>();
+    await mapMapAsync(redisClientsForKeys, async ({ redisClient, keys }) => {
+      const values = await redisClient.mget(keys);
+      for (let i = 0; i < keys.length; i++) {
+        allKeysToValues.set(nullthrows(keys[i]), assertNotUndefined(values[i]));
+      }
+    });
+
+    return originalKeys.map((k) => assertNotUndefined(allKeysToValues.get(k)));
+  }
+  multi(): IRedisTransaction {
+    const accumulatedSets: Map<string, SetInstruction> = new Map();
+    const parentExec = async (): Promise<void> => {
+      const redisClientsForKeys = this.getRedisClientsForKeys([...accumulatedSets.keys()]);
+      const redisClientsForInstructions = mapMap(redisClientsForKeys, ({ keys, redisClient }) => {
+        return {
+          keys,
+          sets: keys.map((k) => assertNotUndefined(accumulatedSets.get(k))),
+          redisClient,
+        };
+      });
+
+      await mapMapAsync(redisClientsForInstructions, async ({ sets, redisClient }) => {
+        let redisTransaction = redisClient.multi();
+        sets.forEach(({ key, value, secondsToken, seconds }) => {
+          redisTransaction = redisTransaction.set(key, value, secondsToken, seconds);
+        });
+        await redisTransaction.exec();
+      });
+    };
+    const ret: IRedisTransaction = {
+      set(key, value, secondsToken, seconds): any {
+        accumulatedSets.set(key, { key, value, secondsToken, seconds });
+        return ret;
+      },
+      async exec(): Promise<any> {
+        return await parentExec();
+      },
+    };
+    return ret;
+  }
+
+  async del(...args: string[]): Promise<void> {
```

Comment:
In a non-test implementation, this should return a number (the sum of the individual del results)